packages/awsfirehose/data_stream/metrics/elasticsearch/ingest_pipeline/default.yml

```
@@ -1,9 +1,47 @@
 ---
 description: Pipeline for rerouting metrics stream from Amazon Data Firehose.
 processors:
+  - dot_expander:
+      field: "*"
+      ignore_failure: true
   - set:
       field: ecs.version
       value: 8.11.0
+  - append:
+      field: "aws.metrics_names"
+      value: [ null ]
+      allow_duplicates: false
+      description: Create an empty field of the object type to temporarily store list of all metrics names
+  - script:
+      lang: painless
+      description: This script adds all metric names from 'aws.*.metrics.*.*' into aws.metrics_names field.
+      if: ctx.aws != null
+      source: >-
+        List metricNames = new ArrayList();
+        if (ctx.aws != null && ctx.aws instanceof Map) {
```

```suggestion
        List metricNames = ctx.aws.metrics_names;
        if (ctx.aws instanceof Map) {
```

?

I'm less certain of the first line change here, we _do_ know that `ctx.aws.metrics_names` exists, but do we know that it is always empty? Does this matter? At the moment (mod the null value that is inserted), any existing name would be clobbered, and we are allocating an array needlessly.

The second line follows from the comment dealing with the removal of the `if` property, but also `ctx.aws != null && ctx.aws instanceof Map` is strictly equivalent to `ctx.aws instanceof Map`.

---

The Firehose receiver will not (should not) emit `aws.metrics_names`, so it doesn't matter if we clobber. Your suggestion seems fine to me. Another option would be:
 - drop the `append` processor
 - keep the `if ctx.aws != null` condition above
 - keep creating the array here
 - add `ignore_missing: true` to the `remove` processor below

---

Yep, that would also be good.

---

Yeah thanks @efd6 and @axw ! Firehose receiver should not have `aws.metrics_names` field but I think what @axw suggested makes sense (and removing a processor is always good). Just made the change! Thank you!